drivers/frequency/adf4371/adf4371.c

```diff
@@ -0,0 +1,820 @@
+
+	return SUCCESS;
+}
+
+/**
+ * Calculate closest possible rate
+ * @param dev - The device structure
+ * @param rate - The desired rate.
+ * @param rounded_rate - The closest possible rate of desired rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_round_rate(struct adf4371_dev *dev, uint64_t rate,
+			       uint64_t *rounded_rate)
+{
+	*rounded_rate = rate;
+
+	return SUCCESS;
+}
+
+/**
+ * Set channel rate.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_set_rate(struct adf4371_dev *dev, uint32_t chan,
+			     uint64_t rate)
+{
+	if (chan >= dev->num_channels)
+		return FAILURE;
+
+	return adf4371_set_freq(dev, rate, chan);
+}
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_init(struct adf4371_dev **device,
+		     const struct adf4371_init_param *init_param)
+{
+	struct adf4371_dev *dev;
+	int32_t ret;
+	uint32_t i;
+
+	dev = (struct adf4371_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret < 0)
+		return ret;
+
+	dev->spi_3wire_en = init_param->spi_3wire_enable;
+
+	dev->clkin_freq = init_param->clkin_frequency;
+
+	dev->differential_ref_clk = init_param->differential_ref_clock;
+
+	if (init_param->loop_filter_inverting)
+		dev->pd_pol = 0;
+	else
+		dev->pd_pol = 1;
+
+	if (init_param->charge_pump_microamp == 0) {
+		/* Default charge pump current is 1.75mA */
+		dev->cp_settings.icp = 1750;
+		dev->cp_settings.regval = 0x04;
+	} else {
+		i = find_closest(init_param->charge_pump_microamp,
+				 adf4371_cp_current_microamp,
+				 ARRAY_SIZE(adf4371_cp_current_microamp));
+		dev->cp_settings.regval = i;
+		dev->cp_settings.icp = adf4371_cp_current_microamp[i];
+	}
+
+	dev->mute_till_lock_en = init_param->mute_till_lock_enable;
+
+	if (init_param->muxout_select > 10)
+		dev->muxout_default_mode = 0; /* Tristate */
+	else
+		dev->muxout_default_mode = init_param->muxout_select;
+
+	if (dev->muxout_default_mode > 0)
+		dev->muxout_en = true;
+
+	dev->muxout_1v8_en = init_param->muxout_level_1v8_enable;
+
+	dev->num_channels = 4;
+
+	for (i = 0; i < init_param->num_channels; i++) {
+		dev->channel_cfg[init_param->channels[i].num].enable = true;
+		dev->channel_cfg[init_param->channels[i].num].freq =
+			init_param->channels[i].power_up_frequency;
+	}
+
+	*device = dev;
+
+	return adf4371_setup(dev);
+}
+
+/**
+ * Remove the device - release resources.
+ * @param device - The device structure.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_remove(struct adf4371_dev *device)
+{
+	int32_t ret;
+
```

Comment:
It should be checked if device is NULL or not.

Reply:
I've fixed this too.